File: sw-pib/src/util/atomic.py
import enum
import functools
import util.variable
from util.variable import *
from util.encrypted_variable import *
from util.resolved_values import *


# Operation input is nothing more than an ordered list of input variables
class OperationInput:

    # Values is a list of Variables (either of lenght 1, 2, or n)
    # depending on whether the input is unary, binary or nary
    def __init__(self, values):
        self.values = values


# Type definition of output variable
OperationOutput = Variable


# Return operation to compute
def get_op(operation_name: str):
    # somehow match does not work even though i am on version 3.10
    if operation_name == "Addition":
        return add
    elif operation_name == "AdditionConst":
        return add_const
    elif operation_name == "AdditionOverN":
        return add_over_n
    elif operation_name == "Absolute":
        return absolute
    elif operation_name == "DefConst":
        return def_const
    elif operation_name == "Division":
        return div
    elif operation_name == "DivisionConstVar":
        return div_const_var
    elif operation_name == "DivisionVarConst":
        return div_var_const
    elif operation_name == "Maxima":
        return maxima
    elif operation_name == "MaximaOverN":
        return max_over_n
    elif operation_name == "Minima":
        return minima
    elif operation_name == "MinimaOverN":
        return min_over_n
    elif operation_name == "Multiplication":
        return mul
    elif operation_name == "MultiplicationConst":
        return mul_const
    elif operation_name == "Squareroot":
        return sqrt
    elif operation_name == "Subtraction":
        return sub
    elif operation_name == "SubtractionConstVar":
        return sub_const_var
    elif operation_name == "SubtractionVarConst":
        return sub_var_const
    elif operation_name == "Power":
        return power
    elif operation_name == "PowerConst":
        return power_const_exp
    elif operation_name == "PowerBaseConst":
        return power_base_const
    else:
        raise Exception("Did not find operation: ", operation_name)


def get_resolved_for_op(atomic, resolved_vals: ResolvedValues, crypto=None):
    # print("Computing ", atomic["op"], atomic["name"],  "with vars", atomic["var"], "and constant", atomic["constant"])
    # NARY Operations
    if atomic["op"] == "Addition" or atomic["op"] == "Subtraction" or atomic["op"] == "Multiplication" or atomic["op"] == "Minima" or atomic["op"] == "Maxima":
        if len(atomic["var"]) == 0:
            raise Exception("No variable for nary operation")
        if atomic["constant"] is not None:
            raise Exception("Nary operation has unused constant")

        values = []
        # extract the required and resolved values
        for var in atomic["var"]:
            values.append(resolved_vals.get(var))

    # BINARY Operations with 2 Variables
    if atomic["op"] == "Division" or atomic["op"] == "Power":
        if len(atomic["var"]) != 2:
            raise Exception("Binary operation expects exactly two inputs")
        if atomic["constant"] is not None:
            raise Exception("Binary operation has unused constant")

        values = [resolved_vals.get(atomic["var"][0]), resolved_vals.get(atomic["var"][1])]

    # BINARY Operations with one var and one constant on second position
    # BINARY Operations with one var and one constant on the first position
    # In both cases: constant is the second position
    if atomic["op"] == "AdditionConst" or atomic["op"] == "SubtractionVarConst" or atomic["op"] == "MultiplicationConst" or atomic["op"] == "DivisionVarConst" or atomic["op"] == "PowerConst" or atomic["op"] == "PowerBaseConst" or atomic["op"] == "SubtractionConstVar" or atomic["op"] == "DivisionConstVar":
        if atomic["constant"] is None:
            raise Exception("Constant and one var expected but no constant given")
        if len(atomic["var"]) != 1:
            raise Exception("Constant and one var expected but", len(atomic["var"]), "vars provided")

        # For plaintext → create a standard variable
        if crypto is None:
            values = [resolved_vals.get(atomic["var"][0]), Variable([atomic["constant"]])]
        # Otherwise create an encrypted cipher and store it alongside
        else:
            const_cipher = create_constant_cipher(crypto, float(atomic["constant"]))
            const_cipher.plain = float(atomic["constant"])
            values = [resolved_vals.get(atomic["var"][0]), const_cipher]

    # UNARY Operation with no constant
    if atomic["op"] == "Squareroot" or atomic["op"] == "Absolute" or atomic["op"] == "AdditionOverN" or atomic["op"] == "MinimaOverN" or atomic["op"] == "MaximaOverN":
        if atomic["constant"] is not None:
            raise Exception("Constant was not expected for unary operation")
        if len(atomic["var"]) != 1:
            raise Exception("Only one var expected but", len(atomic["var"]), "vars provided")

        values = [resolved_vals.get(atomic["var"][0])]

    # Definition of constants → This makes computation much easier
    if atomic["op"] == "DefConst":
        if len(atomic["var"]) != 0:
            raise Exception("No vars expected for DefConst but", len(atomic["var"]), "vars provided")
        if atomic["constant"] is None:
            raise Exception("Constant was expected for DefConst operation")

        # For plaintext → create a standard variable
        if crypto is None:
            values = [Variable([atomic["constant"]])]

        # Otherwise create an encrypted cipher and store it alongside
        else:
            const_cipher = create_constant_cipher(crypto, float(atomic["constant"]))
            const_cipher.plain = float(atomic["constant"])
            values = [const_cipher]

    # print("Got all values on proxy:", values)
    return values


# Constant definition is nothing more than passing the value through
def def_const(input_var: OperationInput) -> OperationOutput:
    # Pass the variable on
    return input_var.values[0]


# Addition of n variables
def add(input_vars: OperationInput) -> OperationOutput:

    # Create zero value (which is turned into ciphertext / variable)
    zero = None

    if isinstance(input_vars.values[0], EncVariable):
        zero = create_constant_cipher(input_vars.values[0].crypto, 0)
    elif isinstance(input_vars.values[0], Variable):
        zero = Variable([0] * len(input_vars.values[0]))
    else:
        raise Exception("Unknown variable type:", type(input_vars.values[0]))

    # Compute sum over all encrypted variables
    result = functools.reduce(operator.add, input_vars.values, zero)

    return result


# Add constant to var
def add_const(input_vars: OperationInput) -> OperationOutput:

    # Both variable types have a length that is checked beforehand
    if len(input_vars.values[0]) != 1:
        raise Exception("Trying to do constant operation with non-scalar!")

    # Operator overloading allows for dynamic choosing of implementation
    result = input_vars.values[0] + input_vars.values[1]

    # Control Messages based on variable type
    if isinstance(input_vars.values[0], EncVariable):
        # print("add_const → encrypted output")
        pass
    elif isinstance(input_vars.values[0], Variable):
        # print("got plain vars:", input_vars.values, "sum is", result)
        pass
    else:
        raise Exception("Unknown variable type:", type(input_vars.values[0]))

    return result


# Scalarize vector
def add_over_n(input_var: OperationInput) -> OperationOutput:

    # For both types this is possible without problem
    result = input_var.values[0].add_over_n()

    if isinstance(input_var.values[0], Variable):
        # print("got vars:", input_var.values, "add_over_n is", result)
        pass
    elif isinstance(input_var.values[0], EncVariable):
        # print("got vars:", input_var.values, "add_over_n is encrypted")
        pass

    return result


# Compute min_over_n
def min_over_n(input_vars: OperationInput) -> OperationOutput:
    # print(type(input_vars.values[0]))
    # print("got vars:", input_vars.values, "min_over_n is", input_vars.values[0].min_over_n())
    return input_vars.values[0].min_over_n()


# Compute max_over_n
def max_over_n(input_vars: OperationInput) -> OperationOutput:
    # print("got vars:", input_vars.values, "max_over_n is", input_vars.values[0].max_over_n())
    return input_vars.values[0].max_over_n()


# Add constant to var
def sub(input_vars: OperationInput) -> OperationOutput:

    # print("SUBTRACTING:", input_vars.values)

    # One argument for subtraction → Negation
    if len(input_vars.values) == 1:

        if isinstance(input_vars.values[0], EncVariable):
            return -input_vars.values[0]
        elif isinstance(input_vars.values[0], Variable):
            vec_len = len(input_vars.values[0])
            negation_vec = Variable([-1] * vec_len)
            return input_vars.values[0] * negation_vec

    minuend = input_vars.values[0]
    subtrahends = input_vars.values[1::1]

    result = functools.reduce(operator.sub, subtrahends, minuend)

    if isinstance(input_vars.values[0], Variable):
        # print("got vars:", input_vars.values, "sub is", result)
        pass
    elif isinstance(input_vars.values[0], EncVariable):
        # print("got vars:", input_vars.values, "sub is encrypted")
        pass

    return result


# Add constant to var (the second variable was already turned into a constant)
# The constant is *always* on second place
def sub_const_var(input_vars: OperationInput) -> OperationOutput:
    if len(input_vars.values[0]) != 1:
        raise Exception("Trying to do constant operation with non-scalar!")
    # print("got vars:", input_vars.values, "sub_const_var is", input_vars.values[1] - input_vars.values[0])
    return input_vars.values[1] - input_vars.values[0]


# The constant is *always* on second place
def sub_var_const(input_vars: OperationInput) -> OperationOutput:
    if len(input_vars.values[0]) != 1:
        raise Exception("Trying to do constant operation with non-scalar!")
    # print("got vars:", input_vars.values, "sub_var_const is", input_vars.values[0] - input_vars.values[1])
    return input_vars.values[0] - input_vars.values[1]


# Multiply two values
def mul(input_vars: OperationInput) -> OperationOutput:

    # Compute sum over all encrypted variables
    if isinstance(input_vars.values[0], EncVariable):
        one = create_constant_cipher(input_vars.values[0].crypto, 1)
    elif isinstance(input_vars.values[0], Variable):
        one = Variable([1] * len(input_vars.values[0]))
    else:
        raise Exception("Unknown variable type:", type(input_vars.values[0]))

    result = input_vars.values[0] * input_vars.values[1]

    return result


# Add constant to var
def mul_const(input_vars: OperationInput) -> OperationOutput:
    if len(input_vars.values[0]) != 1:
        raise Exception("Trying to do constant operation with non-scalar!")
    # print("got vars:", input_vars.values, "mul_const is", input_vars.values[0] * input_vars.values[1])
    return input_vars.values[0] * input_vars.values[1]


# Divide one var through another
def div(input_vars: OperationInput) -> OperationOutput:
    # Otherwise use folding and subtract the whole vector
    # print("got vars:", input_vars.values, "div is", input_vars.values[0] / input_vars.values[1])
    return input_vars.values[0] / input_vars.values[1]


# Divide constant through var (the second variable was already turned into a constant)
# The constant is *always* on second place
def div_const_var(input_vars: OperationInput) -> OperationOutput:
    if len(input_vars.values[0]) != 1:
        raise Exception("Trying to do constant operation with non-scalar!")
    # print("got vars:", input_vars.values, "div_const_var is", input_vars.values[1] / input_vars.values[0])
    return input_vars.values[1] / input_vars.values[0]


# The constant is *always* on second place
def div_var_const(input_vars: OperationInput) -> OperationOutput:
    if len(input_vars.values[0]) != 1:
        raise Exception("Trying to do constant operation with non-scalar!")

    result = None

    if isinstance(input_vars.values[0], EncVariable):
        # Provide true constant and not the ciphertext
        result = input_vars.values[0].div_var_const(input_vars.values[1].plain)
        # print("got vars:", input_vars.values, "div_var_const is encrypted")
    elif isinstance(input_vars.values[0], Variable):
        result = input_vars.values[0] / input_vars.values[1]
        # print("got vars:", input_vars.values, "div_var_const is ", result)
    else:
        raise Exception("Unexpected variable type: ", type(input_vars.values[0]))

    return result


# Take the square root of a variable
def sqrt(input_vars: OperationInput) -> OperationOutput:
    # print("got vars:", input_vars.values, "sqrt is", input_vars.values[0].sqrt())
    return input_vars.values[0].sqrt()


# Power
def power(input_vars: OperationInput) -> OperationOutput:
    result = input_vars.values[0] ** input_vars.values[1]

    if isinstance(input_vars.values[0], EncVariable):
        # print("got vars:", input_vars.values, "div_var_const is encrypted")
        pass
    elif isinstance(input_vars.values[0], Variable):
        # print("got vars:", input_vars.values, "div_var_const is ", result)
        pass
    else:
        raise Exception("Unexpected variable type: ", type(input_vars.values[0]))

    return result


# Power with constant exponent

def power_const_exp(input_vars: OperationInput) -> OperationOutput:

    result = None

    if isinstance(input_vars.values[0], EncVariable):
        result = input_vars.values[0].power_const_exp(input_vars.values[1].plain)
        # print("got vars:", input_vars.values, "power_const_exp is encrypted")
    elif isinstance(input_vars.values[0], Variable):
        result = input_vars.values[0] ** input_vars.values[1]
        # print("got vars:", input_vars.values, "power_const_exp is ", result)
    else:
        raise Exception("Unexpected variable type: ", type(input_vars.values[0]))

    return result


# Power with constant base
# Constant is *always* the second entry!
def power_base_const(input_vars: OperationInput) -> OperationOutput:
    if len(input_vars.values[0]) != 1:
        raise Exception("Trying to do constant operation with non-scalar!")

    result = input_vars.values[1] ** input_vars.values[0]

    if isinstance(input_vars.values[0], EncVariable):
        # print("got vars:", input_vars.values, "power_base_const is encrypted")
        pass
    elif isinstance(input_vars.values[0], Variable):
        # print("got vars:", input_vars.values, "power_base_const is ", result)
        pass
    else:
        raise Exception("Unexpected variable type: ", type(input_vars.values[0]))

    return result


# Power with constant base
def minima(input_vars: OperationInput) -> OperationOutput:
    # print("got vars:", input_vars.values, "minima is", min(input_vars.values))
    return min(input_vars.values)


# Max of n values
def maxima(input_vars: OperationInput) -> OperationOutput:
    # print("got vars:", input_vars.values, "maxima is", max(input_vars.values))
    return max(input_vars.values)


# Compute absolute
def absolute(input_vars: OperationInput) -> OperationOutput:
    # print("got vars:", input_vars.values, "abs is", abs(input_vars.values[0]))
    return abs(input_vars.values[0])


# Apply binary operation in mapping manner
def map_binary_op(first, second, op):
    if len(first) != len(second):
        raise Exception("Vector dimension missmatch: ", len(first), len(second))

    res = list(map(op, zip(first, second)))
    return util.variable.Variable(res)


# Apply unary operation in mapping manner
def map_unary_op(invar, op):
    res = list(map(op, invar))
    return util.variable.Variable(res)

File: hw-pib/impl/benchmark/src/atomic.rs
//! An atomic is the base operation of the analyst's uploaded
//! algorithms. It has a `name` field that specifies the necessarily
//! **unique** identifier for the opration.
//! Each atomic holds an operationType which can be one of in
//! [`OperationType`] specified operations, like addition, mult,...
//! There are var fields that hold the "variables" that are
//! involved in a specific atomic operation. The "variables" refer to
//! dependent atomic operations. This can be one (unary or with the
//! optional [`Const`] constant), two (binary) or $n$ variables ($n$-ary).
//! The dependendent variables have to be resolved first before the
//! computation of the current field can be performed.
//!
//! Additionally the boolean `is_kpi` field states wether the result
//! is important to the end result for companies that participate in the
//! benchmark. Others will be simply discarded.
//!
//! A template of one [`Atomic`] field in YAML:
//!
//!     - name: test_op
//!         op: Addition
//!         is_kpi: true
//!         var:
//!             - three
//!             - two_op
//!             - three
//!             - four
//!
//! The implementation of calc performs the computation of the result for
//! one specific input pair, that is provided (for a company).

use crate::{
    error::BenchmarkingError,
    operation::{
        abs, add, add_const, add_over_n, def_const, div, div_const_var, div_var_const, max,
        max_over_n, min, min_over_n, mul, mul_const, power, power_base_const, sqrt, sub,
        sub_const_var, sub_var_const, OperationInput, OperationOutput, OperationType,
        OperationType::*,
    },
    resolved::ResolvedValues,
    variable::{Const, Variable, VariableID},
};
use serde::{Deserialize, Serialize};

/// Atomic calculation unit
#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct Atomic {
    /// YAML var name
    name: VariableID,
    /// KPI is final result
    is_kpi: bool,
    /// Operation from
    op: OperationType,
    /// input variables for algorithm
    var: Vec<VariableID>,
    #[serde(default)]
    /// Constant operand
    constant: Option<Const>,
}

impl Atomic {
    /// New variable that is required as input but not specified by the analyst as an "atomic"
    pub fn new_required(name: VariableID) -> Self {
        // Required Variables are constant 0 additions
        Atomic {
            name: name.clone(),
            is_kpi: false,
            op: AdditionConst,
            var: Vec::new(),
            constant: Some(0_f64),
        }
    }

    /// make operation valid for computations
    pub fn make_op_valid(&mut self) {
        self.var.push(self.name.clone());
    }

    /// Return name of Atomic
    pub fn name(&self) -> &str {
        &self.name
    }

    /// Return the `OperationType`
    pub fn op(&self) -> OperationType {
        self.op
    }

    /// Returns `true` if this is an kpi
    pub fn is_kpi(&self) -> bool {
        self.is_kpi
    }

    /// Retuns the names of all dependency variables
    pub fn var_ids(&self) -> &[VariableID] {
        &self.var
    }

    /// Retuns the constant value of operation
    pub fn constant(&self) -> &Option<Const> {
        &self.constant
    }

    // For parsing
    pub fn new(
        name: VariableID,
        is_kpi: bool,
        op: OperationType,
        var: Vec<VariableID>,
        constant: Option<Const>,
    ) -> Self {
        Atomic {
            name,
            is_kpi,
            op,
            var,
            constant,
        }
    }

    /// Calculate atomic
    pub fn calc<'a>(&'a self, resolved: &mut ResolvedValues) -> Result<(), BenchmarkingError> {
        // Error handling for calculation
        let throw_computation_error =
            |op_type: OperationType, expected_op_input: &str, receive_op_input: OperationInput| {
                BenchmarkingError::from(format!(
                    "operation {} expects {}, but received {:?}",
                    op_type.to_string(),
                    expected_op_input.to_string(),
                    receive_op_input.to_string()
                ))
            };

        // Operands and operation for computation of atomic
        let input = self.get_resolved_for_op(resolved, false)?;
        let input_op = self.get_map_op();

        let _input_cpy: OperationInput;

        #[cfg(not(feature = "evaluation"))]
        {
            _input_cpy = input.clone();
        }

        // Exhaustive list of operations and their constraints that mitigate panics on runtime
        // When returning, the result is a variable
        let var_res = match self.op {
            Absolute => match input {
                OperationInput::Unary(_) => input_op(input),
                _ => return Err(throw_computation_error(self.op, "Unary", input)),
            },
            Addition => match input {
                OperationInput::NAry(_) => input_op(input),
                _ => return Err(throw_computation_error(self.op, "NAry", input)),
            },
            AdditionConst => match input {
                OperationInput::Binary(_, _) => input_op(input),
                _ => return Err(throw_computation_error(self.op, "Binary", input)),
            },
            AdditionOverN => match input {
                OperationInput::Unary(_) => input_op(input),
                _ => {
                    return Err(throw_computation_error(self.op, "Unary", input));
                }
            },
            MinimaOverN => match input {
                OperationInput::Unary(_) => input_op(input),
                _ => {
                    return Err(throw_computation_error(self.op, "Unary", input));
                }
            },
            MaximaOverN => match input {
                OperationInput::Unary(_) => input_op(input),
                _ => {
                    return Err(throw_computation_error(self.op, "Unary", input));
                }
            },
            Subtraction => match input {
                OperationInput::NAry(_) => input_op(input),
                _ => {
                    return Err(throw_computation_error(self.op, "NAry", input));
                }
            },
            SubtractionConstVar => match input {
                OperationInput::Binary(_, _) => input_op(input),
                _ => {
                    return Err(throw_computation_error(self.op, "Binary", input));
                }
            },
            SubtractionVarConst => match input {
                OperationInput::Binary(_, _) => input_op(input),
                _ => {
                    return Err(throw_computation_error(self.op, "Binary", input));
                }
            },
            Multiplication => match input {
                OperationInput::NAry(_) => input_op(input),
                _ => {
                    return Err(throw_computation_error(self.op, "NAry", input));
                }
            },
            MultiplicationConst => match input {
                OperationInput::Binary(_, _) => input_op(input),
                _ => {
                    return Err(throw_computation_error(self.op, "Binary", input));
                }
            },
            Division => match input {
                OperationInput::Binary(_, _) => input_op(input),
                _ => {
                    return Err(throw_computation_error(self.op, "Binary", input));
                }
            },
            DivisionVarConst => match input {
                OperationInput::Binary(_, _) => input_op(input),
                _ => {
                    return Err(throw_computation_error(self.op, "Binary", input));
                }
            },
            DivisionConstVar => match input {
                OperationInput::Binary(_, _) => input_op(input),
                _ => {
                    return Err(throw_computation_error(self.op, "Binary", input));
                }
            },
            Squareroot => match input {
                OperationInput::Unary(_) => input_op(input),
                _ => {
                    return Err(throw_computation_error(self.op, "Unary", input));
                }
            },
            Power => match input {
                OperationInput::Binary(_, _) => input_op(input),
                _ => {
                    return Err(throw_computation_error(self.op, "Binary", input));
                }
            },
            PowerConst => match input {
                OperationInput::Binary(_, _) => input_op(input),
                _ => {
                    return Err(throw_computation_error(self.op, "Binary", input));
                }
            },
            PowerBaseConst => match input {
                OperationInput::Binary(_, _) => input_op(input),
                _ => {
                    return Err(throw_computation_error(self.op, "Binary", input));
                }
            },
            Minima => match input {
                OperationInput::NAry(_) => input_op(input),
                _ => {
                    return Err(throw_computation_error(self.op, "NAry", input));
                }
            },
            Maxima => match input {
                OperationInput::NAry(_) => input_op(input),
                _ => {
                    return Err(throw_computation_error(self.op, "NAry", input));
                }
            },
            DefConst => match input {
                OperationInput::Unary(_) => input_op(input),
                _ => return Err(throw_computation_error(self.op, "Unary", input)),
            },
        };

        #[cfg(not(feature = "evaluation"))]
        {
            log::info!(
                "[Atomic: {}] [Type: {}] [Input: {:?}] [Output: {:?}]",
                self.name(),
                self.op.to_string(),
                _input_cpy,
                var_res
            );
        }

        // Insert into resolved
        resolved.insert(self.name().to_string(), var_res)?;

        Ok(())
    }

    // Select input values from `resolved_values`. Does some sanity checks and chooses the right amount of numbers
    fn get_resolved_for_op<'a>(
        &'a self,
        resolved_values: &'a ResolvedValues,
        _is_eval: bool,
    ) -> Result<OperationInput, BenchmarkingError> {
        // Prepares the input with the InputType enum. This makes handling easier
        // For example `Addition` input can have arbitrary length.
        // `Power` on the other hand has always two input vars, and `Absolute` has only one input variable.

        // Eval requires another connection mode to peers to communicate with!
        log::info!("{} uses var {:?}", self.name(), &self.var);

        match self.op {
            //
            // NARY OPERATIONS
            //
            Addition | Subtraction | Multiplication | Minima | Maxima => {
                #[cfg(not(feature = "evaluation"))]
                {
                    // Sanity checks: at least one var and no const required
                    if self.var.len() == 0 {
                        return Err(BenchmarkingError::from((self.clone(), "has no variables")));
                    }
                    if let Some(constant) = self.constant {
                        return Err(BenchmarkingError::from((
                            self.clone(),
                            format!("has unused constant {:?}", constant),
                        )));
                    }
                }

                // Extract resolved nary values for further computation
                let mut values = Vec::new();
                for var in &self.var {
                    let val = resolved_values.get(var)?;
                    values.push(val.clone());
                }
                Ok(OperationInput::NAry(values))
            }

            //
            // BINARY OPERATIONS WITH 2 VARIABLES
            //
            Division | Power => {
                #[cfg(not(feature = "evaluation"))]
                {
                    // need exactly 2 operands
                    let len = self.var.len();
                    if len != 2 {
                        return Err(BenchmarkingError::from((
                            self.clone(),
                            format!("binary op expected 2 vars but received {}", len),
                        )));
                    }
                    if let Some(constant) = self.constant {
                        return Err(BenchmarkingError::from((
                            self.clone(),
                            format!("has unused constant {:?}", constant),
                        )));
                    }
                }

                // Extract resolved binary values for further computation
                let var_names = (&self.var[0], &self.var[1]);
                let operands = (
                    resolved_values.get(var_names.0)?,
                    resolved_values.get(var_names.1)?,
                );

                #[cfg(not(feature = "evaluation"))]
                {
                    // Verify against 0-divisions
                    if operands.1.vector().iter().find(|&&x| x == 0_f64).is_some()
                        && self.op == Division
                    {
                        //operands.1 = Variable::new(vec![1.0]);
                        //log::error!("0-Division for {}", var_names.1);
                        return Err(BenchmarkingError::from((
                            self.clone(),
                            format!("has 0-Division for value {}", var_names.1),
                        )));
                    }
                }

                Ok(OperationInput::Binary(
                    operands.0.clone(),
                    operands.1.clone(),
                ))
            }

            //
            // BINARY OPERATIONS WITH ONE VAR AND ONE CONSTANT ON SECOND POSITION
            //
            AdditionConst | SubtractionVarConst | MultiplicationConst | DivisionVarConst
            | PowerConst | PowerBaseConst => {
                #[cfg(not(feature = "evaluation"))]
                {
                    // sanity check: expect one var and one const
                    let len = self.var.len();
                    if len != 1 {
                        return Err(BenchmarkingError::from((
                            self.clone(),
                            format!(
                                "binary op with constant expected 1 var but received {}",
                                len
                            ),
                        )));
                    }
                    if self.constant.is_none() {
                        return Err(BenchmarkingError::from((
                            self.clone(),
                            "is binary op with constant but no constant is provided!",
                        )));
                    }
                }

                // Extract computed values → The constant is always the second operand
                let var_name = &self.var[0];
                let operand0 = resolved_values.get(var_name)?;
                let operand1 = Variable::new(vec![self.constant.unwrap_or_default()]);

                #[cfg(not(feature = "evaluation"))]
                {
                    // Verify against 0-divisions
                    if operand1.vector().iter().find(|&&x| x == 0_f64).is_some()
                        && (self.op == DivisionVarConst || self.op == SubtractionVarConst)
                    {
                        return Err(BenchmarkingError::from((
                            self.clone(),
                            "has 0-Division for provided constant",
                        )));
                    }
                }

                Ok(OperationInput::Binary(operand0.clone(), operand1))
            }

            //
            // BINARY OPERATIONS WITH ONE VAR AND ONE CONSTANT ON FIRST POSITION
            //
            SubtractionConstVar | DivisionConstVar => {
                #[cfg(not(feature = "evaluation"))]
                {
                    // sanity check: expect one var and one const
                    let len = self.var.len();
                    if len != 1 {
                        return Err(BenchmarkingError::from((
                            self.clone(),
                            format!(
                                "binary op with constant expected 1 var but received {}",
                                len
                            ),
                        )));
                    }
                    if self.constant.is_none() {
                        return Err(BenchmarkingError::from((
                            self.clone(),
                            "is binary op with constant but no constant is provided!",
                        )));
                    }
                }

                // Extract computed values → The constant is always the second operand
                // We do error handling here to check whether the divisor is 0.
                let var_name = &self.var[0];
                let operand0 = resolved_values.get(var_name)?;
                let operand1 = Variable::new(vec![self.constant.unwrap_or_default()]);

                #[cfg(not(feature = "evaluation"))]
                {
                    // Verify against 0-divisions
                    if operand0.vector().iter().find(|&&x| x == 0_f64).is_some()
                        && (self.op == DivisionVarConst || self.op == SubtractionVarConst)
                    {
                        //operand1 = Variable::new(vec![1.0]);
                        //log::error!("0-Division for {:?}", operand0);
                        return Err(BenchmarkingError::from((
                            self.clone(),
                            format!(
                                "has 0-Division for provided variable operand: {:?}",
                                operand0
                            ),
                        )));
                    }
                }

                Ok(OperationInput::Binary(operand0.clone(), operand1))
            }

            //
            // UNARY OPERATION WITH NO CONSTANT
            //
            Squareroot | Absolute | AdditionOverN | MinimaOverN | MaximaOverN => {
                #[cfg(not(feature = "evaluation"))]
                {
                    // sanity checks: one operand only
                    let len = self.var.len().clone();
                    if len != 1 {
                        return Err(BenchmarkingError::from((
                            self.clone(),
                            format!("binary op expected 2 vars but received {}", len).as_str(),
                        )));
                    }
                    if let Some(constant) = self.constant {
                        return Err(BenchmarkingError::from((
                            self.clone(),
                            format!("has unused constant {:?}", constant),
                        )));
                    }
                }

                // Extract resolved values
                let var_name = &self.var[0];
                let n = resolved_values.get(var_name)?;
                Ok(OperationInput::Unary(n.clone()))
            }

            //
            // DEFINITION OF CONSTANTS
            //
            DefConst => {
                #[cfg(not(feature = "evaluation"))]
                {
                    // sanity checks: no variable should exist
                    let len = self.var.len().clone();
                    if len != 0 {
                        return Err(BenchmarkingError::from((
                            self.clone(),
                            format!("constant definition expected 0 vars but received {}", len)
                                .as_str(),
                        )));
                    }
                    if self.constant.is_none() {
                        return Err(BenchmarkingError::from((
                            self.clone(),
                            format!("has no constant"),
                        )));
                    }
                }

                // Extract resolved values
                let operand0 = Variable::new(vec![self.constant.unwrap_or_default().into()]);
                Ok(OperationInput::Unary(operand0))
            }
        }
    }

    #[allow(unused_parens)]
    /// Map the operation type to the function that performs it
    fn get_map_op(&self) -> (fn(OperationInput) -> OperationOutput) {
        match self.op() {
            Addition => add,
            AdditionConst => add_const,
            AdditionOverN => add_over_n,
            MinimaOverN => min_over_n,
            MaximaOverN => max_over_n,
            Subtraction => sub,
            SubtractionConstVar => sub_const_var,
            SubtractionVarConst => sub_var_const,
            Multiplication => mul,
            MultiplicationConst => mul_const,
            Division => div,
            DivisionConstVar => div_const_var,
            DivisionVarConst => div_var_const,
            Squareroot => sqrt,
            Power | PowerConst => power,
            PowerBaseConst => power_base_const,
            Minima => min,
            Maxima => max,
            Absolute => abs,
            DefConst => def_const,
        }
    }
}
